File: HeapSort/Python/TernaryHeapSort.py

import time #To measure time#

#Heapifying Process#
def HeapsortA(A, n, parent):

  #Create number of nodes#
  leftnode=parent*3+1
  middlenode=parent*3+2
  rightnode=parent*3+3


  #Check if left node is larger than parent#
  if leftnode<n and A[leftnode]> A[parent]:
    tmp=leftnode
  else:
    tmp=parent

  #Check if middle node is larger than parent and left node#
  if(middlenode<n and A[middlenode]> A[tmp]):
    tmp=middlenode
  
  #Check if right node is larger than parent, right and left node#
  if(rightnode<n and A[rightnode]> A[tmp]):
    tmp=rightnode


  #If either children is larger, perform swap and rebalance with swap parent#
  if(tmp != parent):
    A[parent], A[tmp] = A[tmp], A[parent]
    HeapsortA(A,n,tmp)


#Structure of Heap Sort#
def Heapsort(A, n):

  #Calculate half the element of the array#
  size=n//3-1

  #Part A: Heapify#
  for size in range(size, -1, -1):
    HeapsortA(A, n, size)
 
 #Part B: Removing each element and rebalancing#
  for last in range(n-1, 0, -1):
    arr[last], arr[0] = arr[0], arr[last] 
    HeapsortA(A, last, 0)


 #The data set# 
arr=[573, -187, -400, -356, -100, 440, 355, -116, -9, 229, -5, 969, -179, 894, -664, 205, -753, -132, 271, 441, 976, -931, -67, -247, -382, 889, 496, -487, -998, -154, -573, 864, 962, -572, -676, -534, 478, -691, -555, -386, -868, -766, -660, 197, 742, 933, -943, 887, -201, -898, 971, -607, -260, 40, -148, 353, -449, 93, 52, -548, 174, 250, 96, -174, -231, 233, 286, 70, -372, -960, 358, 22, -981, -554, -562, 558, 746, 490, -621, -822, -86, 696, -642, -93, -873, 411, 619, -926, 780, -203, 420, -553, -523, -447, -68, 786, 645, 674, -383, 448

]

#Start stopwatch#
start = time.time_ns()
n=len(arr)
Heapsort(arr,n)

#Stop the stopwatch#
end = time.time_ns()-start

#Print array to confirm#
print(end)
print("ns")
print("The Sorted-Array is: ", arr) 

File: HeapSort/C++/TernaryHeapSort.cpp
#include <iostream>
using namespace std;

//To measure running time//
#include <chrono>
#include <unistd.h>

void printArray(int A[], int size){//Print Array//
  for (auto i = 0; i < size; i++){
    cout << A[i] << " ";}
}


//Part A: Check if parent is the largest among its children//
void HeapSortA(int A[],int n, int parent){
  int tmp;
  //Calculating the child position//
  int leftnode=parent*3+1;
  int middlenode=parent*3+2;
  int rightnode=parent*3+3;

  //Check if Left child is larger than parent//
  if(leftnode<n && A[leftnode]> A[parent]){
    tmp=leftnode;
  }else{
    tmp=parent;
  }

  //Check if middle child is larger than left child or parent//
  if(middlenode<n && A[middlenode]> A[tmp]){
    tmp=middlenode;
  }

  //Check if right child is larger than left, middle child or parent//
  if(rightnode<n && A[rightnode]> A[tmp]){
    tmp=rightnode;
  }

  //Perform swap and repeat for swaped element to acheive balance heap binary tree again//
  if(tmp!=parent){
    swap(A[parent],A[tmp]);
    HeapSortA(A,n,tmp);
  }
}


void HeapSort(int A[],int n){

  //Part A: Heapify//
  for (int size=n/3-1; size>=0; size--){
    HeapSortA(A,n,size);
  }


  //Part B: Removing and rebalancing//
  for(int last=n-1; last>0; last--){
    swap(A[0],A[last]);
    HeapSortA(A,last,0);
  }

}



int main() {
//Data Set//
int A[] = {100, 99, 98, 97, 96, 95, 94, 93, 92, 91,90, 89, 88, 87, 86, 85, 84, 83, 82, 81,80, 79, 78, 77, 76, 75, 74, 73, 72, 71,70, 69, 68, 67, 66, 65, 64, 63, 62, 61,60, 59, 58, 57, 56, 55, 54, 53, 52, 51,50, 49, 48, 47, 46, 45, 44, 43, 42, 41,40, 39, 38, 37, 36, 35, 34, 33, 32, 31,30, 29, 28, 27, 26, 25, 24, 23, 22, 21,20, 19, 18, 17, 16, 15, 14, 13, 12, 11,10, 9, 8, 7, 6, 5, 4, 3, 2, 1};

//Start the stopwatch//
auto start = chrono::steady_clock::now();
auto arr_size = sizeof(A) / sizeof(A[0]); 

//Sorting function//
HeapSort(A,arr_size);

//Stop the stopwatch
auto end = chrono::steady_clock::now();

cout << "Elapsed time in nanoseconds: "
    << chrono::duration_cast<chrono::nanoseconds>(end - start).count()<< " ns" << endl;

//Print out array//
cout << "\nSorted array is \n";
  printArray(A, arr_size);
}
